jb/fftw/ut_time_delay_estimator_many.cpp

```diff
@@ -0,0 +1,335 @@
+#include <jb/fftw/tde_result.hpp>
+#include <jb/fftw/time_delay_estimator_many.hpp>
+#include <jb/testing/check_tde_result_close_enough.hpp>
+#include <jb/testing/create_square_timeseries.hpp>
+#include <jb/testing/create_triangle_timeseries.hpp>
+#include <jb/testing/delay_timeseries.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include <chrono>
+
+/**
+ * @test Verify that we can create and use a simple time delay estimator.
+ * dimension 3
+ */
+BOOST_AUTO_TEST_CASE(fftw_time_delay_estimator_many_3_dim_tde) {
+  int const nsamples = 1 << 15;
+  int const S = 20;
+  int const V = 4;
+  int const delay = 2500;
+  int const tol = 2;
+
+  using array_type = jb::fftw::aligned_multi_array<float, 3>;
+  using tested_type = jb::fftw::time_delay_estimator_many<array_type>;
+  using confidence_type = typename tested_type::confidence_type;
+  using argmax_type = typename tested_type::argmax_type;
+
+  array_type a(boost::extents[S][V][nsamples]);
+  array_type b(boost::extents[S][V][nsamples]);
+  confidence_type confidence(a);
+  // expected confidence result to compare within tolerance tol
+  confidence_type expected_confidence(a);
+
+  argmax_type argmax(a);
+  // expected argmax result to compare within tolerance tol
+  argmax_type expected_argmax(a);
+
+  using sum2_type = jb::fftw::tde_result<array_type, float>;
+  // sum of square value of timeseries
+  sum2_type sum2(b);
+
+  // construct the tested FTE
+  tested_type tested(a, b);
+
+  // fill the timeseries with triangle or square
+  jb::testing::create_triangle_timeseries(nsamples, b);
+
+  // a = delay shift timeseries b
+  int count = 0;
+  for (int i = 0; i != S; ++i) {
+    for (int j = 0; j != V; ++j, ++count) {
+      float acc_sum2 = 0;
+      for (int k = 0; k != nsamples; ++k) {
+        a[i][j][k] = b[i][j][(k + delay) % nsamples];
+        acc_sum2 += a[i][j][k] * a[i][j][k];
+      }
+      sum2[count] = acc_sum2;
+      expected_argmax[count] = static_cast<std::size_t>(delay);
+      expected_confidence[count] = static_cast<float>(nsamples);
+    }
+  }
+
+  // run the TDE...
+  (void)tested.estimate_delay(a, b, sum2, confidence, argmax);
+  // check argmax within delay +/- tol
+  BOOST_CHECK_MESSAGE(
+      jb::testing::tde_result_close_enough(argmax, expected_argmax, tol),
+      "argmax is not within tolerance("
+          << tol << ", argmax[0]=" << argmax[0]
+          << ", expected_argmax[0]=" << expected_argmax[0]);
+  // check confidence within nsamples +/- tol
```

Comment:
nsamples is `1<<15`, while you are testing for `10*tol`.  The comment and the code do not agree.  I think I documented somewhere the expected error of a cross-correlation based on the precision of FFTs and so forth.... yup:

https://github.com/coryan/jaybeams/blob/master/jb/fftw/ut_plan_many.cpp#L23